### DISCUSSION TITLE
Undo Redo manager with mobx-utils `deepObserve`

### DISCUSSION THREAD

**Discussion by steveruizok** [mobxjs/mobx#3281] [Show and tell]
Hey all, I saw that undo redo has come up a few times in the issues. I was looking for a solution and couldn't find one, so here's what I came up with: https://codesandbox.io/s/mobx-undo-redo-v2-ofty1 . I'd love to hear any thoughts on improving it, or if you have any similar solutions that I could look at!

**Comment by steveruizok** [comment]
Ah I've run into a snag: like observe, deepObserve does not respect transactions, meaning that a change that effects multiple properties will trigger multiple "undos". We can debounce the deepObserve's listener, but this means the undo/redo will no longer be synchronous!

Here's a sandbox showing what this looks like: https://codesandbox.io/s/mobx-undo-redo-v3-vx2oo

**Reply by urugator** [reply]
https://github.com/mobxjs/mobx/discussions/3106#discussioncomment-1305960

**Comment by steveruizok** [comment]
Thanks to @urugator, here's a new version that runs synchronously: https://codesandbox.io/s/mobx-undo-redo-v4-mwyi6

**Reply by kubk** [reply]
@steveruizok Looks good 👍  Are you going to make a separate library out of it?

**Reply by steveruizok** [reply]
There are probably a few more bugs hiding in there, but if it's solid then sure!
